sql/hive/src/main/java/org/apache/hadoop/hive/ql/io/orc/SparkVectorizedOrcRecordReader.java

```diff
@@ -0,0 +1,189 @@
+        case PRIMITIVE:
+          {
+            PrimitiveTypeInfo primitiveTypeInfo =
+              (PrimitiveTypeInfo) ((PrimitiveObjectInspector)inspector).getTypeInfo();
+            switch(primitiveTypeInfo.getPrimitiveCategory()) {
+              case BOOLEAN:
+              case BYTE:
+              case SHORT:
+              case INT:
+              case LONG:
+              case DATE:
+              case INTERVAL_YEAR_MONTH:
+                return new LongColumnVector(VectorizedRowBatch.DEFAULT_SIZE);
+              case FLOAT:
+              case DOUBLE:
+                return new DoubleColumnVector(VectorizedRowBatch.DEFAULT_SIZE);
+              case BINARY:
+              case STRING:
+              case CHAR:
+              case VARCHAR:
+                BytesColumnVector column = new BytesColumnVector(VectorizedRowBatch.DEFAULT_SIZE);
+                column.initBuffer();
+                return column;
+              case DECIMAL:
+                DecimalTypeInfo tInfo = (DecimalTypeInfo) primitiveTypeInfo;
+                return new DecimalColumnVector(VectorizedRowBatch.DEFAULT_SIZE,
+                    tInfo.precision(), tInfo.scale());
+              default:
+                throw new RuntimeException("Vectorizaton is not supported for datatype:"
+                    + primitiveTypeInfo.getPrimitiveCategory());
+            }
+          }
+        default:
+          throw new RuntimeException("Vectorization is not supported for datatype:"
+              + inspector.getCategory());
+      }
+    }
+
+    /**
+     * Walk through the object inspector and add column vectors
+     *
+     * @param oi StructObjectInspector
+     * @param cvList ColumnVectors are populated in this list
+     */
+    private void allocateColumnVector(StructObjectInspector oi,
+        List<ColumnVector> cvList) {
+      if (cvList == null) {
+        throw new RuntimeException("Null columnvector list");
+      }
+      if (oi == null) {
+        return;
+      }
+      final List<? extends StructField> fields = oi.getAllStructFieldRefs();
+      for(StructField field : fields) {
+        ObjectInspector fieldObjectInspector = field.getFieldObjectInspector();
+        cvList.add(createColumnVector(fieldObjectInspector));
+      }
+    }
+
+    /**
+     * Create VectorizedRowBatch from ObjectInspector
+     *
+     * @param oi StructObjectInspector
+     * @return VectorizedRowBatch
+     */
+    private VectorizedRowBatch constructVectorizedRowBatch(
+        StructObjectInspector oi) {
+      final List<ColumnVector> cvList = new LinkedList<ColumnVector>();
+      allocateColumnVector(oi, cvList);
+      final VectorizedRowBatch result = new VectorizedRowBatch(cvList.size());
+      int i = 0;
+      for(ColumnVector cv : cvList) {
+        result.cols[i++] = cv;
+      }
+      return result;
+    }
+
+    @Override
+    public boolean next(NullWritable key, VectorizedRowBatch value) throws IOException {
```

Comment:
Do you mean a batch is empty and then its next batch has some rows? I think it is not possible.